Out of tree AMD GPU support

mgsharm opened the pull request:
**Description of changes:**

This PR adds AMD GPU DKMS driver support for Bottlerocket kernel 6.12, enabling users to run newer AMD GPU hardware with the latest out-of-tree drivers. It also includes systemd services to ensure proper driver initialization before dependent services start.

**Dependencies**

⚠️ Blocked by: https://github.com/bottlerocket-os/bottlerocket-kernel-kit/pull/325

This PR requires AMD GPU firmware support to be merged first. The firmware files are necessary for the AMD GPU drivers to initialize properly. 

**Changes**

1. Disable In-Tree AMD GPU Driver

Why: Out-of-tree DKMS drivers typically support newer GPU hardware faster than in-tree drivers and receive updates independently of kernel releases.

What changed:
 - Disabled CONFIG_DRM_AMDGPU in both x86_64 and aarch64 kernel configs
 - Updated kernel patch 1006 to explicitly select DRM helper modules (DRM_DISPLAY_HDCP_HELPER and DRM_DISPLAY_HDMI_HELPER) that were previously bundled with the in-tree driver but are required as standalone modules for DKMS builds
 - Eliminates driver conflicts by ensuring exclusive hardware access

2. Add AMD GPU DKMS Kernel Module

What: New kmod-6.12-amdgpu package that builds AMD GPU drivers from upstream DKMS source.

Key features:
  - Sources drivers from AMD's official DKMS repository (version 30.20)
  - Dynamic kernel version detection - automatically adapts to kernel updates without hardcoded dependencies
  - Full cross-compilation support integrated with Bottlerocket's build toolchain
  - GPG signature verification of upstream driver packages
  - Includes all required kernel modules: amdgpu, amdkcl, amdttm, amddrm_*, amd-sched, amdxcp

**Testing**

  Build Validation

  - ✅ Package compiles successfully in Bottlerocket's cross-compilation environment
  - ✅ Cross-compilation verified: x86_64 → aarch64 and aarch64 → x86_64 builds both successful
  - ✅ AMI creation completed and instance boots correctly

  Hardware Validation

  Tested on AMD GPU instance with 8 GPUs (device ID 75a3 - MI300X):

  Driver Initialization

  All GPUs initialized successfully:
```bash  
bash-5.1# dmesg | grep -E 'amdgpu.*initialized'
  [   40.434923] amdgpu 0000:51:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
  [   42.672485] amdgpu 0000:51:00.0: amdgpu: SMU is initialized successfully!
  [   43.326766] amdgpu 0000:52:00.0: amdgpu: RAS INFO: ras initialized successfully, hardware ability[65bff] ras_mask[65bff]
  [   45.583236] amdgpu 0000:52:00.0: amdgpu: SMU is initialized successfully!
  # ... (8 GPUs total, all initialized successfully)
```
```bash
  All GPUs detected by PCI:
  bash-5.1# lspci | grep -i 75a3 | wc -l
  8
```

```bash
bash-5.1# lspci -vvv | grep -i 75a3 -A 2 -B 2
        Kernel driver in use: ena

51:00.0 Processing accelerators: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Subsystem: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Physical Slot: 79-1
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
--
        Kernel driver in use: amdgpu

52:00.0 Processing accelerators: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Subsystem: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Physical Slot: 80-1
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
--
        Kernel driver in use: pcieport

62:00.0 Processing accelerators: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Subsystem: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Physical Slot: 93-1
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
--
        Kernel driver in use: amdgpu

63:00.0 Processing accelerators: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Subsystem: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Physical Slot: 94-1
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
--
        Kernel driver in use: pcieport

73:00.0 Processing accelerators: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Subsystem: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Physical Slot: 107-1
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
--
        Kernel driver in use: amdgpu

74:00.0 Processing accelerators: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Subsystem: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Physical Slot: 108-1
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
--
        Kernel driver in use: pcieport

84:00.0 Processing accelerators: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Subsystem: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Physical Slot: 121-1
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
--
        Kernel driver in use: amdgpu

85:00.0 Processing accelerators: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Subsystem: Advanced Micro Devices, Inc. [AMD/ATI] Device 75a3
        Physical Slot: 122-1
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
```

  Package Installation

```bash
  bash-5.1# cat /var/lib/bottlerocket/inventory/application.json | grep -A 10 "kmod-6.12-amdgpu"
  {
    "Name": "kmod-6.12-amdgpu",
    "Publisher": "bottlerocket-kernel-kit",
    "Version": "30.10",
    "Release": "1.1763654568.d2bc9648.br1",
    "Epoch": "0",
    "InstalledTime": "2025-11-20T20:59:51Z",
    "ApplicationType": "Unspecified",
    "Architecture": "x86_64",
    "Url": "https://www.amd.com/",
    "Summary": "AMD GPU drivers for the 6.12 kernel"
  }
```

Terms of Contribution

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

yeazelm commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+%global kernel_major 6.12
```
I think we can move this into the kernel-6.12 build so its just included with the rest of the kernel built in one shot.

yeazelm commented on packages/linux-firmware-amd/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+[package]
+name = "linux-firmware-amd"
```
I had expected this to just be a sub-package of `linux-firmware` instead of a new top level package. Can we move this logic into the existing package?

yeazelm commented on packages/linux-firmware-amd/linux-firmware-amd.spec:
```diff
@@ -0,0 +1,43 @@
+
+%files
+%dir %{fwdir}
+%{fwdir}/*
```
I don't think we want everything, I think there is a specific set of firmware we want that is only amdgpu related.

yeazelm commented on packages/kmod-6.12-amdgpu/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+[[package.metadata.build-package.external-files]]
+url = "https://repo.radeon.com/amdgpu/30.10.2/el/10/main/x86_64/amdgpu-dkms-6.14.14-2226257.el10.noarch.rpm"
+sha512 = "ffe054d3673f729786a32cf8a67d69b942348eb12d800073a7efc873fcba8a760485a003ed1bfb5273eda5f7d514263c80a97243371d36bf69ea986137f8c41e"
+force-upstream = true
```
I think we can skip this and treat it like a normal source.

bcressey replied:
```suggestion
```

arnaldo2792 commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+pushd amdgpu
+
+# Build using the DKMS Makefile with static configuration
+make modules   KERNELVER=%{_cross_kver}   kernel_build_dir=%{_cross_usrsrc}/kernels/6.12   CC=%{_cross_target}-gcc   ARCH=%{_cross_karch}   CROSS_COMPILE=%{_cross_target}-   EXTRA_CFLAGS=-DPACKAGE_VERSION=\\\"%{version}\\\"
```
Why do you need the triple `\` at the end?

arnaldo2792 commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+Version: 30.10
+Release: 1%{?dist}
+Summary: AMD GPU drivers for the 6.12 kernel
+License: GPL-2.0 WITH Linux-syscall-note
```
The license doesn't match the license in the `RPM`.

mgsharm replied:
Will update them.

arnaldo2792 commented on packages/kmod-6.12-amdgpu/COPYING:
```diff
@@ -0,0 +1,20 @@
+The Linux Kernel is provided under:
```
This is not the correct license, the RPM includes a license already:

```
./share/doc/amdgpu-dkms/LICENSE
```

arnaldo2792 commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+%{summary}.
+
+%prep
+# Extract DKMS source from RPM
```
This is not compiling with `DKMS` so you could just say extract driver srouces.

mgsharm replied:
Makes sense.

arnaldo2792 commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+pushd amdgpu
+
+# Configure DKMS
+KERNELVER=%{_cross_kver} amd/dkms/configure --with-linux=%{kernel_sources}
```
You could use the `cross_configure` macro and likely avoid the flags in the `make modules` call.

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+pushd amdgpu
+
+# Configure DKMS
+KERNELVER=%{_cross_kver} amd/dkms/configure --with-linux=%{kernel_sources}
```
`configure` scripts should not be run during `%prep`, they go in `%build`

mgsharm replied:
Yes, will move to `%build` .

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+%install
+# Install license file
+install -d %{buildroot}%{_cross_licensedir}/%{name}
+install -p -m 0644 %{S:1} %{buildroot}%{_cross_licensedir}/%{name}/
```
This license should come from `usr/share/doc/amdgpu-dkms/LICENSE` in the RPM, it shouldn't be included as a source file.

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+
+# Install AMD GPU kernel modules to correct location
+install -d %{buildroot}%{_cross_kmoddir}/kernel/drivers/gpu/drm/amd/amdgpu
+install -p -m 644 amdgpu/amd/amdkcl/amdkcl.ko %{buildroot}%{_cross_kmoddir}/kernel/drivers/gpu/drm/amd/amdgpu/
```
these should use the `%{_ko}` macro:
```suggestion
install -p -m 644 amdgpu/amd/amdkcl/amdkcl.%{_ko} %{buildroot}%{_cross_kmoddir}/kernel/drivers/gpu/drm/amd/amdgpu/
```

mgsharm replied:
Will use `%{_ko}` macro.

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,72 @@
+install -d %{buildroot}%{_cross_kmoddir}/kernel/drivers/gpu/drm/amd/amdgpu
+install -p -m 644 amdgpu/amd/amdkcl/amdkcl.ko %{buildroot}%{_cross_kmoddir}/kernel/drivers/gpu/drm/amd/amdgpu/
+install -p -m 644 amdgpu/ttm/amdttm.ko %{buildroot}%{_cross_kmoddir}/kernel/drivers/gpu/drm/amd/amdgpu/
+install -p -m 644 amdgpu/amddrm_ttm_helper.ko %{buildroot}%{_cross_kmoddir}/kernel/drivers/gpu/drm/amd/amdgpu/
```
A for loop over the `.ko` files would be more readable if all these kmods are going into the same directory.

bcressey reviewed (commented):
When aarch64 fails, that's a sign that cross-compiling might not be set up correctly. It's not always true, but it's the theory you should work hard to disprove.

Based on that I'd say you should get the aarch64 build at least to the point where it fails from some source-level compatibility. Otherwise I don't trust that the x86_64 build is doing what we think, meaning the configure / make logic may not be correct.

arnaldo2792 commented on packages/kernel-6.12/kernel-6.12.spec:
```diff
@@ -18,6 +19,8 @@ Source2: https://yum.repos.neuron.amazonaws.com/aws-neuronx-dkms-2.21.37.0.noarc
 Source3: https://yum.repos.neuron.amazonaws.com/aws-neuronx-dkms-2.24.7.0.noarch.rpm
 Source4: gpgkey-00FA2C1079260870A76D2C285749CAD8646D9185.asc
 Source5: https://efa-installer.amazonaws.com/aws-efa-installer-1.44.0.tar.gz
+# AMD GPU DKMS driver for x86_64
```
We can live without this comment :smile: 
```suggestion
```

bcressey commented on packages/kernel-6.12/2000-amd-dkms-Disable-configure-drm_get_panel_min_brightness_quirk.patch:
```diff
@@ -0,0 +1,39 @@
+diff --git a/amd/dkms/configure b/amd/dkms/configure
+index 0c166c5..12b7aba 100755
+--- a/amd/dkms/configure
++++ b/amd/dkms/configure
```
Since this is a generated file, the preferred approach is to patch `amd/dkms/aclocal.m4` to remove this macro, then run `autoreconf -fi` prior to regenerate `configure` prior to running it.

mgsharm committed:
kernel-6.12: disable in-tree AMD GPU driver

Disable CONFIG_DRM_AMDGPU and remove related kernel modules from
packaging to allow DKMS AMD driver to take precedence.

Signed-off-by: Gaurav Sharma <mgsharm@amazon.com>

mgsharm committed:
kmod-6.12-amdgpu: add out of tree AMD GPU driver

- Add new kmod-6.12-amdgpu package with out of tree AMD GPU driver support
- Update kernel 1006 patch to include additional DRM helpers for AMD GPU

Signed-off-by: Gaurav Sharma <mgsharm@amazon.com>

bcressey commented on packages/kmod-6.12-amdgpu/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+publish = false
+build = "../build.rs"
+
+[package.metadata.build-package]
+package-name = "kmod-6.12-amdgpu"
```
```suggestion
[package.metadata.build-package]
package-name = "kmod-6.12-amdgpu"
releases-url = "https://repo.radeon.com/amdgpu/"
```

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,81 @@
+Release: 1%{?dist}
+Summary: AMD GPU drivers for the 6.12 kernel
+License: GPL-2.0 WITH Linux-syscall-note
+URL: https://www.amd.com/
```
```suggestion
URL: https://repo.radeon.com/amdgpu/
```

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,81 @@
+Version: 30.10
+Release: 1%{?dist}
+Summary: AMD GPU drivers for the 6.12 kernel
+License: GPL-2.0 WITH Linux-syscall-note
```
```suggestion
License: MIT AND GPL-2.0-only AND (GPL-2.0-only WITH Linux-syscall-note) AND GPL-2.0-or-later AND (GPL-2.0-or-later WITH Linux-syscall-note) AND LGPL-2.0-or-later
```

Derived from this list:
```
❯ rg -N -I 'SPDX-License-Identifier'|sed 's,[/\*#],,g'|awk '{$1=$1};1'|sort -u
SPDX-License-Identifier: BSD-3-Clause OR GPL-2.0
SPDX-License-Identifier: GPL-2.0
SPDX-License-Identifier: GPL-2.0+
SPDX-License-Identifier: GPL-2.0 AND MIT
SPDX-License-Identifier: GPL-2.0-only
SPDX-License-Identifier: GPL-2.0-or-later
SPDX-License-Identifier: GPL-2.0 OR MIT
SPDX-License-Identifier: GPL-2.0+ WITH Linux-syscall-note
SPDX-License-Identifier: ((GPL-2.0 WITH Linux-syscall-note) OR BSD-3-Clause)
SPDX-License-Identifier: (GPL-2.0 WITH Linux-syscall-note) OR MIT
SPDX-License-Identifier: LGPL-2.0+
SPDX-License-Identifier: MIT
```

`BSD-3-Clause` omitted because we might as well pick `GPL-2.0-only` for that case.

mgsharm replied:
Will update this

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,81 @@
+    scheduler/amd-sched \
+    amd/amdxcp/amdxcp \
+    amd/amdgpu/amdgpu; do
+  install -p -m 644 amdgpu/${module}.%{_ko} %{buildroot}%{_cross_kmoddir}/kernel/drivers/gpu/drm/amd/amdgpu/
```
Seems like this path will be more consistent with the NVIDIA drivers:
```suggestion
  install -p -m 644 amdgpu/${module}.%{_ko} %{buildroot}%{_cross_kmoddir}/kernel/drivers/extra/gpu/drm/amd/amdgpu
```

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,81 @@
+  INSTALL_HDR_PATH="%{buildroot}%{_cross_prefix}" \
+  INSTALL_MOD_PATH="%{buildroot}%{_cross_prefix}" \
+  INSTALL_MOD_STRIP=1 \
+  %{nil}}
```
nit: let's move this just before %build

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,81 @@
+    amddrm_exec \
+    scheduler/amd-sched \
+    amd/amdxcp/amdxcp \
+    amd/amdgpu/amdgpu; do
```
if these are all the drivers, we can use:
```
find amdgpu -type f -name "*.%{_ko}" -print -exec install -p -m 0644 '{}' 
  %{buildroot}%{_cross_kmoddir}/kernel/drivers/extra/gpu/drm/amd/amdgpu
```

and rely on the explicit inclusions in `%files` to flag any changes

mgsharm replied:
Sounds good.

yeazelm commented on packages/kmod-6.12-amdgpu/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+path = "../packages.rs"
+
+[[package.metadata.build-package.external-files]]
+url = "https://repo.radeon.com/amdgpu/30.10.2/el/10/main/x86_64/amdgpu-dkms-6.14.14-2226257.el10.noarch.rpm"
```
Can we try the newer version: https://repo.radeon.com/amdgpu/30.20/el/10/main/x86_64/?

mgsharm replied:
Yes, will update to use `30.20`.

yeazelm commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,74 @@
+  %{nil}}
+
+%build
+# Build AMD GPU DKMS driver
```
```suggestion
# Build AMD GPU driver
```

yeazelm commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,74 @@
+# Build AMD GPU DKMS driver
+pushd %{_builddir}/amdgpu
+# Configure DKMS driver
+pushd amd/dkms
```
nit: this should probably just be one line
```suggestion
pushd %{_builddir}/amdgpu/amd/dkms
```

yeazelm commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,74 @@
+  ARCH="%{_cross_karch}" \
+  --with-linux=%{kernel_sources} \
+  %{nil}
+popd
```
If you take the change above, then you'll need to `pushd %{_builddir}/amdgpu` but this feels a bit more direct than the nesting before.

bcressey commented on packages/kmod-6.12-amdgpu/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+path = "../packages.rs"
+
+[[package.metadata.build-package.external-files]]
+url = "https://repo.radeon.com/amdgpu/30.20/el/10/main/x86_64/amdgpu-dkms-6.16.6-2238411.el10.noarch.rpm"
```
This RPM is signed, so we should validate the signature.

mgsharm replied:
Yes, I'll add them.

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,74 @@
+%install
+# Install AMD GPU kernel modules to correct location
+install -d %{buildroot}%{_cross_kmoddir}/kernel/drivers/extra/gpu/drm/amd/amdgpu
+find amdgpu -type f -name "*.%{_ko}" -print -exec install -p -m 0644 '{}' %{buildroot}%{_cross_kmoddir}/kernel/drivers/extra/gpu/drm/amd/amdgpu/ \;
```
nit: split line for readability:
```suggestion
find amdgpu -type f -name "*.%{_ko}" -print -exec install -p -m 0644 '{}' \
  %{buildroot}%{_cross_kmoddir}/kernel/drivers/extra/gpu/drm/amd/amdgpu/ \;
```

bcressey commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,74 @@
+%global kernel_sources %{_cross_usrsrc}/kernels/6.12
+%define _kernel_version %(cat %{kernel_sources}/include/config/kernel.release)
+%global _cross_kmoddir %{_cross_libdir}/modules/%{_kernel_version}
```
define `amdgpu_kmoddir` to cut down on repetition of the install path:
```suggestion
%global _cross_kmoddir %{_cross_libdir}/modules/%{_kernel_version}
%global amdgpu_kmoddir %{_cross_kmoddir}/kernel/drivers/extra/gpu/drm/amd/amdgpu
```

mgsharm commented:
Added linux-firmware dependency for amd gpu to initialize - `Requires: %{_cross_os}linux-firmware-amdgpu` .

arnaldo2792 commented on packages/kmod-6.12-amdgpu/kmod-6.12-amdgpu.spec:
```diff
@@ -0,0 +1,80 @@
+%global kernel_sources %{_cross_usrsrc}/kernels/6.12
+%define _kernel_version %(cat %{kernel_sources}/include/config/kernel.release)
+%global _cross_kmoddir %{_cross_libdir}/modules/%{_kernel_version}
+%global amdgpu_kmoddir %{_cross_kmoddir}/kernel/drivers/extra/gpu/drm/amd/amdgpu
```
Please use `updates` instead of `extra`:

```suggestion
%global amdgpu_kmoddir %{_cross_kmoddir}/kernel/drivers/updates/gpu/drm/amd/amdgpu
```
